std/internal/math/biguintcore.d

```diff
@@ -345,35 +346,47 @@ public:
     }
 
     // return false if invalid character found
-    bool fromHexString(const(char)[] s) pure nothrow @safe
+    bool fromHexString(Range)(Range s) if (
+        isBidirectionalRange!Range && isSomeChar!(ElementType!Range))
     {
+        import std.range : walkLength;
+
         //Strip leading zeros
-        int firstNonZero = 0;
-        while ((firstNonZero < s.length - 1) &&
-            (s[firstNonZero]=='0' || s[firstNonZero]=='_'))
-        {
-                ++firstNonZero;
-        }
-        auto len = (s.length - firstNonZero + 15)/4;
-        auto tmp = new BigDigit[len+1];
-        uint part = 0;
-        uint sofar = 0;
-        uint partcount = 0;
-        assert(s.length>0);
-        for (ptrdiff_t i = s.length - 1; i>=firstNonZero; --i)
-        {
-            assert(i>=0);
-            char c = s[i];
-            if (s[i]=='_') continue;
-            uint x = (c>='0' && c<='9') ? c - '0'
-                   : (c>='A' && c<='F') ? c - 'A' + 10
-                   : (c>='a' && c<='f') ? c - 'a' + 10
+        size_t first_non_zero;
+        while (!s.empty && (s.front == '0' || s.front == '_'))
+        {
+            ++first_non_zero;
+            s.popFront;
+        }
+
+        if (s.empty)
+        {
+            data = ZERO;
+            return true;
+        }
+
+        auto len = (s.save.walkLength - first_non_zero + 15) / 4;
```

Comment:
This line is the cause of your bug.  In the `while` loop just before this, you already consumed all the leading zeroes, so `walkLength` no longer includes the zeroes in the count. However, you still subtract by `first_non_zero`. When the latter happens to be longer than the length of the remaining digits, a `size_t` underflow happens and wraps around to a very large value (to be precise, 4611686018427387903), which the following line then tries to allocate a buffer for.  So the out-of-memory error should not be surprising. :-)

The fix is trivial: there is no need to subtract `first_non_zero` at all, since you have already discarded those zeroes in the preceding `while` loop. Remove that, and everything should work as expected.

(Aren't you glad there's a unittest that just happens to have a test case that triggered this bug? It would have slipped in unnoticed otherwise...)